All: Throwing from destructors

Early signs of this are seen in #449 and #577, but I think its a broader problem than those issues make it seem. This is not a clang-specific problem but rather goes down to the C++ standard.

Basically, because destructors need to be safe to call during the stack unwinding as a response to an exception being throw, they should not themselves be able to throw exceptions. 

It is worthwhile to read libMesh's experience with this problem using GCC 6.1, see libMesh/libmesh#1073. This quote from that discussion sums up the relevant language in the various C++ standards as well as the sensible solution of taking out any throw statements from destructors:

> No, not undefined behavior. In the C++03 standard, sections 15.2 and 15.5.1 make it clear that an exception which escapes a destructor results in an immediate call to std::terminate().
>
> In C++11 it's even a bit looser, as best as I can tell: if you override the default "noexcept" specification on your destructor, and your destructor isn't currently being called as part of stack unwinding during another exception throw, and the object being destroyed isn't static or TLS, then the exception is allowed to propagate and be handled normally; otherwise you get std::terminate().
>
> Nitpicking aside, though, after a little more thought my vote is to just remove the libmesh_parallel_only() macro call here. The hypothetical advantage of maybe being able to more easily debug MPI sync failures at LibMeshInit destruction time isn't worth the definite annoyance of a false positive warning or of trying to work around the warning.

Here is a warning thrown by GCC 6.1 on Teuchos:

```
/projects/AppComp/alexa/nightly/cee-compute011/src/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
     throw Exception(omsgstr); \
                            ^
/projects/AppComp/alexa/nightly/cee-compute011/src/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:305:3: note: in expansion of macro ‘TEUCHOS_TEST_FOR_EXCEPTION’
   TEUCHOS_TEST_FOR_EXCEPTION(throw_exception_test, std::logic_error, "Error!")
   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/projects/AppComp/alexa/nightly/cee-compute011/src/Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.cpp:655:5: note: in expansion of macro ‘TEUCHOS_TEST_FOR_EXCEPT’
     TEUCHOS_TEST_FOR_EXCEPT(0==rcp_node_list());^M
     ^~~~~~~~~~~~~~~~~~~~~~~
/projects/AppComp/alexa/nightly/cee-compute011/src/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: note: in C++11 destructors default to noexcept
     throw Exception(omsgstr); \
                            ^
/projects/AppComp/alexa/nightly/cee-compute011/src/Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:305:3: note: in expansion of macro ‘TEUCHOS_TEST_FOR_EXCEPTION’
   TEUCHOS_TEST_FOR_EXCEPTION(throw_exception_test, std::logic_error, "Error!")
   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/projects/AppComp/alexa/nightly/cee-compute011/src/Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.cpp:655:5: note: in expansion of macro ‘TEUCHOS_TEST_FOR_EXCEPT’
     TEUCHOS_TEST_FOR_EXCEPT(0==rcp_node_list());^M
     ^~~~~~~~~~~~~~~~~~~~~~~
```

The way I see it, here are the options Teuchos has:
1. Remove `TEST_FOR_EXCEPT` calls from destructors.
2. Replace those calls with the equivalent of `assert()`.
3. Try to override the `noexcept` specification. Note that doing this will not guarantee a proper unwinding exception in all cases.

#449 and #577 seem to indicate that there are Teuchos tests which rely on exception throwing from destructors to cause unwinding. Given that this is only standards-conforming with C++11 and even then only under certain special conditions, those tests should be looked at carefully to see if they actually conform to the restrictions mentioned in the standard and if not then expected behavior should be re-evaluated and tests changed.

@trilinos/teuchos @MicheldeMessieres @bartlettroscoe @mhoemmen

Using the following cmake command:

cmake 1> cmake.log 2> cmake.err \
    -D Trilinos_ENABLE_DEVELOPMENT_MODE=OFF \
    -D CMAKE_BUILD_TYPE:STRING=RELEASE \
    -D Trilinos_ENABLE_Sacado:BOOL=ON \
    -D Sacado_ENABLE_TESTS:BOOL=OFF \
    -D CMAKE_INSTALL_PREFIX:PATH=$HOME/prefix/sacado \
    -D Trilinos_INSTALL_LIB_DIR=$HOME/prefix/sacado/$libdir \
    ../

I tried both the master and trilinos-release-12-8-branch branches with and still got the warnings. Here is the first one:

... snip ...
Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
     throw Exception(omsgstr); \
... snip ...

The attached shell script install_sacado.txt (below) has two settings at the top:
    trilinos_repo="$HOME/repo/trilinos.git"
   trilinos_version='trilinos-release-12-8-branch'
The first you will have to change to the location of your trilinos repository.

The final output when I run
    bash install_sacado.txt
on my system, is:
[100%] Built target sacado
/home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
/home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
install_sacado.txt: terminate warnings in make command stderr output


[install_sacado.txt](https://github.com/trilinos/Trilinos/files/1039932/install_sacado.txt)

https://github.com/bartlettroscoe

I ran the following commands:
gcc --version | head -1
cd /home/bradbell/repo/trilinos.git/build/packages/teuchos/core/src
make VERBOSE=1 Teuchos_RCPNode.o

and got the following output:
gcc (GCC) 6.3.1 20161221 (Red Hat 6.3.1-1)
make[1]: Entering directory '/home/bradbell/repo/trilinos.git/build'
Building CXX object packages/teuchos/core/src/CMakeFiles/teuchoscore.dir/Teuchos_RCPNode.cpp.o
cd /home/bradbell/repo/trilinos.git/build/packages/teuchos/core/src && /usr/bin/c++   -DTEUCHOSCORE_LIB_EXPORTS_MODE -I/home/bradbell/repo/trilinos.git/build -I/home/bradbell/repo/trilinos.git/build/packages/teuchos/core/src -I/home/bradbell/repo/trilinos.git/packages/teuchos/core/src -I/home/bradbell/repo/trilinos.git/build/packages/kokkos/core/src -I/home/bradbell/repo/trilinos.git/packages/kokkos/core/src  -std=c++11 -O3   -o CMakeFiles/teuchoscore.dir/Teuchos_RCPNode.cpp.o -c /home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_RCPNode.cpp
In file included from /home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_Assert.hpp:46:0,
                 from /home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_any.hpp:51,
                 from /home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_RCPNode.hpp:53,
                 from /home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_RCPNode.cpp:42:
/home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_RCPNode.cpp: In destructor ‘Teuchos::ActiveRCPNodesSetup::~ActiveRCPNodesSetup()’:
/home/bradbell/repo/trilinos.git/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
     throw Exception(omsgstr); \
  ... snip ...

Thanks Ross, sounds great.  Just wanted to add that for a package like ROL, resolving this issue would help us focus on important compiler warnings.  As we sit downstream of pretty much everything in Trilinos, there are thousands of these warnings when compiling ROL, and any other potential warnings are completely swamped by

```
/home/dridzal/development/ROL-Trilinos/packages/teuchos/core/src/Teuchos_TestForException.hpp:183:28: warning: throw will always call terminate() [-Wterminate]
     throw Exception(omsgstr); \
                            ^
/home/dridzal/development/ROL-Trilinos/packages/teuchos/core/src/Teuchos_RCPNode.hpp:522:7: note: in expansion of macro ‘TEUCHOS_TEST_FOR_EXCEPTION’
       TEUCHOS_TEST_FOR_EXCEPTION( ptr_!=0, std::logic_error,
```

--- a/packages/belos/epetra/src/BelosEpetraAdapter.cpp
+++ b/packages/belos/epetra/src/BelosEpetraAdapter.cpp
@@ -273,7 +273,7 @@ namespace Belos {
 
 	  const int info = 
 	    const_cast<Epetra_Operator &>(Op_).SetUseTranspose (originalTransposeFlag_);
-	  TEUCHOS_TEST_FOR_EXCEPTION(info != 0, EpetraOpFailure,
+	  TEUCHOS_TEST_FOR_TERMINATION(info != 0,
 			     "Resetting the original "
 			     "transpose flag value of the Epetra_Operator failed, "
 			     "returning a nonzero error code of " << info << ".  "
@@ -301,8 +301,7 @@ namespace Belos {
 	// have the right value on the second call.  This would make the
 	// resulting exception message confusing.
 	const bool finalTransposeFlag = Op_.UseTranspose ();
-	TEUCHOS_TEST_FOR_EXCEPTION(originalTransposeFlag_ != finalTransposeFlag,
-			   std::logic_error,
+	TEUCHOS_TEST_FOR_TERMINATION(originalTransposeFlag_ != finalTransposeFlag,
 			   "Belos::OperatorTraits::Apply: The value of the "
 			   "Epetra_Operator's transpose flag changed unexpectedly!"
 			   "  The original value at the top of this method was "

--- a/packages/ml/src/Utils/ml_MultiLevelPreconditioner.h
+++ b/packages/ml/src/Utils/ml_MultiLevelPreconditioner.h
@@ -1051,7 +1051,7 @@ template <class T> void MLVec<T>::wrap(T *start, T *end, bool okToChangePtr,
 
 template <class T> MLVec<T>::~MLVec()              { 
    if (data_ != NULL) {
-      TEUCHOS_TEST_FOR_EXCEPTION(size_ == 0,std::logic_error,
+      TEUCHOS_TEST_FOR_TERMINATION(size_ == 0,
       "MLVec error, data pointer should be null for 0 length vectors\n");
       if (okToChangePtr_ && (freeType_ == useFree))   ML_free(data_); 
       if (okToChangePtr_ && (freeType_ == useDelete)) delete data_;

--- a/packages/teuchos/core/src/Teuchos_RCPNode.cpp
+++ b/packages/teuchos/core/src/Teuchos_RCPNode.cpp
@@ -691,7 +691,7 @@ ActiveRCPNodesSetup::~ActiveRCPNodesSetup()
     std::cerr << "\nPrint active nodes!\n";
 #endif // TEUCHOS_SHOW_ACTIVE_REFCOUNTPTR_NODE_TRACE
     std::cout << std::flush;
-    TEUCHOS_TEST_FOR_EXCEPT(0==rcp_node_list());
+    TEUCHOS_TEST_FOR_TERMINATION(nullptr==rcp_node_list(), "rcp_node_list() is null in ~ActiveRCPNodesSetup");
     RCPNodeTracer::RCPNodeStatistics rcpNodeStatistics =
       RCPNodeTracer::getRCPNodeStatistics();
     if (rcpNodeStatistics.maxNumRCPNodes

--- a/packages/teuchos/core/src/Teuchos_RCPNode.hpp
+++ b/packages/teuchos/core/src/Teuchos_RCPNode.hpp
@@ -586,7 +586,7 @@ public:
   ~RCPNodeTmpl()
     {
 #ifdef TEUCHOS_DEBUG
-      TEUCHOS_TEST_FOR_EXCEPTION( ptr_!=0, std::logic_error,
+      TEUCHOS_TEST_FOR_TERMINATION( ptr_!=0,
         "Error, the underlying object must be explicitly deleted before deleting"
         " the node object!" );
 #endif

--- a/packages/teuchos/core/src/Teuchos_TestForException.cpp
+++ b/packages/teuchos/core/src/Teuchos_TestForException.cpp
@@ -41,6 +41,7 @@
 
 #include "Teuchos_TestForException.hpp"
 
+#include <iostream>
 
 //
 // ToDo: Make these functions thread-safe!
@@ -108,4 +109,9 @@ void Teuchos::TestForException_setEnableStacktrace(bool enableStrackTrace)
 bool Teuchos::TestForException_getEnableStacktrace()
 {
   return loc_enableStackTrace();
+}
+
+void Teuchos::TestForTermination_terminate(const std::string &msg) {
+  std::cerr << msg;
+  std::terminate();
 }

--- a/packages/teuchos/core/src/Teuchos_TestForException.hpp
+++ b/packages/teuchos/core/src/Teuchos_TestForException.hpp
@@ -361,6 +361,21 @@ catch(const std::exception &except) { \
   throw std::runtime_error(omsg.str()); \
 }
 
+/** \brief This macro is to be used instead of
+ * <tt>TEUCHOS_TEST_FOR_EXCEPTION()</tt> to report an error in situations
+ * where an exception can't be throw (like in an destructor).
+ *
+ * \param terminate_test [in] See <tt>TEUCHOS_TEST_FOR_EXCEPTION()</tt>.
+ *
+ * \param msg [in] See <tt>TEUCHOS_TEST_FOR_EXCEPTION()</tt>.
+ *
+ * If the termination test evaluates to <tt>true</tt>, then
+ * <tt>std::terminate()</tt> is called (which should bring down an entire
+ * multi-process MPI program even if only one process calls
+ * <tt>std::terminate()</tt> with most MPI implementation).
+ *
+ * \ingroup TestForException_grp
+ */
 #define TEUCHOS_TEST_FOR_TERMINATION(terminate_test, msg) \
 { \
   const bool call_terminate = (terminate_test); \

--- a/packages/teuchos/core/src/Teuchos_TestForException.hpp
+++ b/packages/teuchos/core/src/Teuchos_TestForException.hpp
@@ -75,6 +75,9 @@ TEUCHOSCORE_LIB_DLL_EXPORT void TestForException_setEnableStacktrace(bool enable
  * exceptions are thrown. */
 TEUCHOSCORE_LIB_DLL_EXPORT bool TestForException_getEnableStacktrace();
 
+/** \brief Prints the message to std::cerr and calls std::terminate. */
+TEUCHOSCORE_LIB_DLL_EXPORT [[noreturn]] void TestForTermination_terminate(const std::string &msg);
+
 
 } // namespace Teuchos
 
@@ -358,7 +361,19 @@ catch(const std::exception &except) { \
   throw std::runtime_error(omsg.str()); \
 }
 
-
-
+#define TEUCHOS_TEST_FOR_TERMINATION(terminate_test, msg) \
+{ \
+  const bool call_terminate = (terminate_test); \
+  if (call_terminate) { \
+    std::ostringstream omsg; \
+    omsg \
+      << __FILE__ << ":" << __LINE__ << ":\n\n" \
+      << "Terminate test that evaluated to true: "#terminate_test \
+      << "\n\n" \
+      << msg << "\n\n"; \
+    auto str = omsg.str(); \
+    Teuchos::TestForTermination_terminate(str); \
+  } \
+}
 
 #endif // TEUCHOS_TEST_FOR_EXCEPTION_H

--- a/packages/teuchos/core/src/Teuchos_TestForException.hpp
+++ b/packages/teuchos/core/src/Teuchos_TestForException.hpp
@@ -372,7 +372,7 @@ catch(const std::exception &except) { \
  * If the termination test evaluates to <tt>true</tt>, then
  * <tt>std::terminate()</tt> is called (which should bring down an entire
  * multi-process MPI program even if only one process calls
- * <tt>std::terminate()</tt> with most MPI implementation).
+ * <tt>std::terminate()</tt> with most MPI implementations).
  *
  * \ingroup TestForException_grp
  */

--- a/packages/teuchos/core/src/Teuchos_Workspace.cpp
+++ b/packages/teuchos/core/src/Teuchos_Workspace.cpp
@@ -160,8 +160,8 @@ RawWorkspace::~RawWorkspace()
   }
   else {
     if(workspace_store_) {
-      TEUCHOS_TEST_FOR_EXCEPTION(
-        workspace_store_->curr_ws_ptr_ != workspace_end_, std::logic_error
+      TEUCHOS_TEST_FOR_TERMINATION(
+        workspace_store_->curr_ws_ptr_ != workspace_end_
         ,"RawWorkspace::~RawWorkspace(...): Error, "
         "Invalid usage of RawWorkspace class, corrupted WorspaceStore object!" );
       workspace_store_->curr_ws_ptr_ = workspace_begin_;

--- a/packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
+++ b/packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
@@ -405,6 +405,17 @@ FOREACH(PID RANGE ${testTeuchosTestForTermination_NP_m_1})
 ENDFOREACH()
 
 
+# Test call to terminate on bad destructor call
+IF (${PARENT_PACKAGE_NAME}_ENABLE_DEBUG)
+  TRIBITS_ADD_EXECUTABLE_AND_TEST( testTeuchosRCPNodeImplDestructTerminate
+    SOURCES testTeuchosRCPNodeImplDestructTerminate.cpp
+    NUM_MPI_PROCS 1
+    PASS_REGULAR_EXPRESSION
+      "Error, the underlying object must be explicitly deleted before deleting the node object"
+    )
+ENDIF()
+
+
 #
 # RCP Beginners Guide example code
 #

--- a/packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
+++ b/packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
@@ -373,7 +373,7 @@ ENDIF()
 #
 # Test TEUCHOS_TEST_FOR_TERMINATION()
 #
-# This aborts the program so the only way to test it is to run a program.
+# This macro aborts the program so the only way to test it is to run a program.
 #
 
 TRIBITS_ADD_EXECUTABLE_AND_TEST( testTeuchosTestForTermination
@@ -401,7 +401,8 @@ FOREACH(PID RANGE ${testTeuchosTestForTermination_NP_m_1})
       "Bingo, we are terminating on procid == terminate_on_procid = ${PID}"
     )
   # NOTE: If each of these tests does not terminate right away, then that is a
-  # sign that std::terminate() on a single process does not
+  # sign that std::terminate() on a single MPI process does not cause all other
+  # processes to also terminate.
 ENDFOREACH()
 
 

--- a/packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
+++ b/packages/teuchos/core/test/MemoryManagement/CMakeLists.txt
@@ -371,6 +371,41 @@ ENDIF()
 
 
 #
+# Test TEUCHOS_TEST_FOR_TERMINATION()
+#
+# This aborts the program so the only way to test it is to run a program.
+#
+
+TRIBITS_ADD_EXECUTABLE_AND_TEST( testTeuchosTestForTermination
+  SOURCES testTeuchosTestForTermination.cpp
+  NUM_MPI_PROCS 1
+  PASS_REGULAR_EXPRESSION
+    "Bingo, we are terminating on procid == terminate_on_procid = 0"
+  )
+
+
+SET(testTeuchosTestForTermination_NP 4)
+MATH(EXPR testTeuchosTestForTermination_NP_m_1
+  "${testTeuchosTestForTermination_NP}-1")
+FOREACH(PID RANGE ${testTeuchosTestForTermination_NP_m_1})
+  TRIBITS_ADD_ADVANCED_TEST(
+    testTeuchosTestForTermination_${PID}_MPI_${testTeuchosTestForTermination_NP}
+    COMM mpi
+    TEST_0
+      MESSAGE "Call TEUCHOS_TEST_FOR_TERMINATION(true) only on PID=${PID} which should terminate all processes"
+      EXEC testTeuchosTestForTermination
+      ARGS --terminate-on-procid=${PID}
+      NUM_MPI_PROCS ${testTeuchosTestForTermination_NP}
+      PASS_REGULAR_EXPRESSION_ALL
+      "p=${PID}: .*/testTeuchosTestForTermination.cpp"
+      "Bingo, we are terminating on procid == terminate_on_procid = ${PID}"
+    )
+  # NOTE: If each of these tests does not terminate right away, then that is a
+  # sign that std::terminate() on a single process does not
+ENDFOREACH()
+
+
+#
 # RCP Beginners Guide example code
 #
 SET(EXAMPLE_RAW_CPP_POINTERS_CPP

--- a/packages/teuchos/core/test/MemoryManagement/testTeuchosRCPNodeImplDestructTerminate.cpp
+++ b/packages/teuchos/core/test/MemoryManagement/testTeuchosRCPNodeImplDestructTerminate.cpp
@@ -1,0 +1,63 @@
+/*
+// @HEADER
+// ***********************************************************************
+//
+//                    Teuchos: Common Tools Package
+//                 Copyright (2004) Sandia Corporation
+//
+// Under terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+// license for use of this work by or on behalf of the U.S. Government.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+// @HEADER
+*/
+
+#include "Teuchos_GlobalMPISession.hpp"
+#include "Teuchos_RCPNode.hpp"
+
+int main(int argc, char* argv[])
+{
+
+  Teuchos::GlobalMPISession mpiSession(&argc,&argv);
+
+  {
+    int obj = 1;
+    using Dealloc_T = Teuchos::DeallocDelete<int>;
+    Teuchos::RCPNodeTmpl<int, Teuchos::DeallocDelete<int> >
+      rcpNode(&obj, Dealloc_T(), false);
+  }
+  // When the above destructor executes it should terminate the program with
+  // an error message!
+
+  return 1; // Will never be called!
+
+}

--- a/packages/teuchos/core/test/MemoryManagement/testTeuchosTestForTermination.cpp
+++ b/packages/teuchos/core/test/MemoryManagement/testTeuchosTestForTermination.cpp
@@ -1,3 +1,4 @@
+/*
 // @HEADER
 // ***********************************************************************
 //
@@ -38,84 +39,29 @@
 //
 // ***********************************************************************
 // @HEADER
+*/
 
-#include "Teuchos_TestForException.hpp"
+#include "Teuchos_CommandLineProcessor.hpp"
 #include "Teuchos_GlobalMPISession.hpp"
+#include "Teuchos_TestForException.hpp"
 
-#include <iostream>
-
-//
-// ToDo: Make these functions thread-safe!
-//
-
-
-namespace {
-
-
-int throwNumber = 0;
-
-
-bool& loc_enableStackTrace()
+int main(int argc, char* argv[])
 {
-  static bool static_enableStackTrace =
-#ifdef HAVE_TEUCHOS_DEFAULT_STACKTRACE
-    true
-#else
-    false
-#endif
-    ;
-  return static_enableStackTrace;
-}
 
+  using Teuchos::GlobalMPISession;
+  GlobalMPISession mpiSession(&argc,&argv);
 
-} // namespace
+  Teuchos::CommandLineProcessor clp;
+  int terminate_on_procid = 0;
+  clp.setOption("terminate-on-procid", &terminate_on_procid);
+  (void)clp.parse(argc, argv);
 
+  TEUCHOS_TEST_FOR_TERMINATION(
+    GlobalMPISession::getRank() == terminate_on_procid,
+    "Bingo, we are terminating on procid == "
+    "terminate_on_procid = "<<GlobalMPISession::getRank()<<"!"
+    );
 
-void Teuchos::TestForException_incrThrowNumber()
-{
-  ++throwNumber;
-}
+  return 1; // Will never be called!
 
-
-int Teuchos::TestForException_getThrowNumber()
-{
-  return throwNumber;
-}
-
-
-void Teuchos::TestForException_break( const std::string &errorMsg )
-{
-  size_t break_on_me;
-  break_on_me = errorMsg.length(); // Use errMsg to avoid compiler warning.
-  if (break_on_me)
-    ; // Avoid 'used var' warning
-  // Above is just some statement for the debugger to break on.  Note: now is
-  // a good time to examine the stack trace and look at the error message in
-  // 'errorMsg' to see what happened.  In GDB just type 'where' or you can go
-  // up by typing 'up' and moving up in the stack trace to see where you are
-  // and how you got to this point in the code where you are throwning this
-  // exception!  Typing in a 'p errorMsg' will show you what the error message
-  // is.  Also, you should consider adding a conditional breakpoint in this
-  // function based on a specific value of 'throwNumber' if the exception you
-  // want to examine is not the first exception thrown.
-}
-
-
-void Teuchos::TestForException_setEnableStacktrace(bool enableStrackTrace)
-{
-  loc_enableStackTrace() = enableStrackTrace;
-}
-
-
-bool Teuchos::TestForException_getEnableStacktrace()
-{
-  return loc_enableStackTrace();
-}
-
-void Teuchos::TestForTermination_terminate(const std::string &msg) {
-  if (GlobalMPISession::getNProc() > 1) {
-    std::cerr << "p="<<GlobalMPISession::getRank()<<": ";
-  }
-  std::cerr << msg << "\n";
-  std::terminate();
 }

--- a/packages/teuchos/doc/build_docs
+++ b/packages/teuchos/doc/build_docs
@@ -23,12 +23,20 @@ echo
 echo "Generating RefCountPtrBeginnersGuideSAND.pdf file ..."
 echo
 
-pushd ../../../doc/RefCountPtr/beginners; make pdf ; popd
-cp ../../../doc/RefCountPtr/beginners/RefCountPtrBeginnersGuideSAND.pdf html/.
+if [ -e ../../../doc/RefCountPtr/beginners ] ; then
+  pushd ../../../doc/RefCountPtr/beginners; make pdf ; popd
+  cp ../../../doc/RefCountPtr/beginners/RefCountPtrBeginnersGuideSAND.pdf html/.
+else
+  echo "Skipping building RefCountPtrBeginnersGuideSAND.pdf because source directory does not exist!"
+fi
 
 echo
 echo "Generating TeuchosMemoryManagementSAND.pdf file ..."
 echo
 
-pushd ../../../doc/TeuchosMemoryManagementSAND; make; popd
-cp ../../../doc/TeuchosMemoryManagementSAND/TeuchosMemoryManagementSAND.pdf html/.
+if [ -e ../../../doc/TeuchosMemoryManagementSAND ] ; then
+  pushd ../../../doc/TeuchosMemoryManagementSAND; make; popd
+  cp ../../../doc/TeuchosMemoryManagementSAND/TeuchosMemoryManagementSAND.pdf html/.
+else
+  echo "Skipping building TeuchosMemoryManagementSAND.pdf because directory does not exist!"
+fi

--- a/packages/teuchos/doc/index.doc
+++ b/packages/teuchos/doc/index.doc
@@ -215,14 +215,14 @@ contains basic, general-purpose utilities.
 
   <ul>
 
-  <li> <em>Testing for Exceptions</em>: The TEUCHOS_TEST_FOR_EXCEPTION() macro accepts
-  a logical test, an exception handler, and a message.  It then throws
-  an "intelligent" exception informing the user of the file, line, and
-  message where the exception was thrown.  The macro
-  TEUCHOS_TEST_FOR_EXCEPT() is a shorter macro that just accepts a
-  logic test and is therefore easier to write.  Please use
-  TEUCHOS_TEST_FOR_EXCEPT(!test) as a safer, more informative
-  alternative to assert(test).</li>
+  <li> <em>Testing for Exceptions</em>: The TEUCHOS_TEST_FOR_EXCEPTION() macro
+  accepts a logical test, an exception handler, and a message.  It then throws
+  an "intelligent" exception informing the user of the file, line, and message
+  where the exception was thrown.  The macro TEUCHOS_TEST_FOR_EXCEPT() is a
+  shorter macro that just accepts a logic test and is therefore easier to
+  write.  Please use TEUCHOS_TEST_FOR_EXCEPT(!test) as a safer, more
+  informative alternative to assert(test).  For related error handling
+  utilties, see \ref TestForException_grp</li>
 
   <li> <em>Stack Tracing</em>: When a debug build is enabled
   (CMAKE_BUILD_TYPE=DEBUG OR Trilinos_ENABLE_DEBUG=ON) and when using GCC and

--- a/packages/thyra/core/src/support/operator_vector/client_support/Thyra_DetachedVectorView.hpp
+++ b/packages/thyra/core/src/support/operator_vector/client_support/Thyra_DetachedVectorView.hpp
@@ -342,7 +342,7 @@ public:
   ~DetachedVectorView()
     {
       if( sv_s_.stride() != sv_.stride() ) {
-        TEUCHOS_TEST_FOR_EXCEPT_MSG(true, "I don't think non-unit stride has ever been tested!");
+        Teuchos::TestForTermination_terminate("I don't think non-unit stride has ever been tested!");
         //Teuchos_Ordinal i; Scalar *sv_v; const Scalar *values;
         //for (
         //  sv_v = sv_s_.values().get(), values = sv_.values().get(), i=0;

